README.md
# pipette
![test](https://github.com/IQTLabs/pipette/workflows/test/badge.svg) [![codecov](https://codecov.io/gh/IQTLabs/pipette/branch/main/graph/badge.svg)](https://codecov.io/gh/IQTLabs/pipette) ![buildx](https://github.com/IQTLabs/pipette/workflows/buildx/badge.svg)

## An SDN/NFV coprocessor controller.
Pipette is a tool that allows users to multiplex SDN coprocessing by implementing transparent L3 NAT. Pipette does this by creating a virtual network behind your coprocessor port and then acting as the SDN controller of that network. Packets are seamlessly switched to their appropriate destination using [Ryu](https://osrg.github.io/ryu/).

## Usage

 1. If an OVS container is not already present, start OVS: `docker-compose -f docker-compose-ovs.yml up -d`. If you have OVS running already (eg, you are running dovesnap), skip this step.
 1. Start pipette: `COPROINT=<ethX> VLANS=<VLANs> NFVIPS=<NFVIPs> OF=<OF TCP port> ID=0 docker-compose -p 0 up -d` (see Configuration section - example `COPROINT=eth1 VLANS=2 NFVIPS=10.10.0.1/16 OF=6699 ID=0 docker-compose -p 0 up -d`.
 1. Start fake services listening on the NFVIP address assigned to the fake interface (Eg, IP of `fake0.2` for VLAN 2 - pipette manages this interface and assigns the NFVIP). Fake services do not have to be in Docker.
 1. When finished, `docker-compose down`.
 1. If you want to run pipette on multiple interfaces, specify different ID and project number as well as interfaces and NFVIPs (the same VLANs can be coprocessed differently on different interfaces - example `COPROINT=eth2 VLANS=2 NFVIPS=10.20.0.1/16 OF=6799 ID=1 docker-compose -p 1 up -d`)

### Configuration
#### Required
 1. COPROINT - the interface that will receive coprocessed packets
 1. VLANS - Space delimitted list of VLANs to coprocess from, must match a VLAN in FAUCET ACL rule
 1. NFVIPS - IPs to send coprocessed packets to. Must be a /16 if IPv4, or /96 if IPv6. There must be the same number of IPs in the list as VLANs.
 1. OF - OpenFlow TCP port for pipette to use


CHANGELOG.md
### [1.11](https://github.com/IQTLabs/pipette/compare/v1.10...v1.11) (2022-03-23)

Upgrade os-ken, alpine, OVS, docker, pytype.

### [1.10](https://github.com/IQTLabs/pipette/compare/v1.9...v1.10) (2021-11-09)

Upgrade pytype, OVS, docker, eventlet.
Switch to OSKen from Ryu.

### [1.9](https://github.com/IQTLabs/pipette/compare/v1.8...v1.9) (2021-06-17)

Upgrade pytype, OVS, docker, alpine.

### [1.8](https://github.com/IQTLabs/pipette/compare/v1.7...v1.8) (2021-05-04)

Upgrade pytype, codecov.

### [1.7](https://github.com/IQTLabs/pipette/compare/v1.6...v1.7) (2021-03-05)

Upgrade to OVS 2.14.2, pytype and and eventlet.

### [1.6](https://github.com/IQTLabs/pipette/compare/v1.5...v1.6) (2021-01-29)

Support multiple pipette instances.

### [1.5](https://github.com/IQTLabs/pipette/compare/v1.4...v1.5) (2021-01-20)

Add tc processing support.

### [1.4](https://github.com/IQTLabs/pipette/compare/v1.3...v1.4) (2021-01-20)

Allow starting pipette from docker-compose, simplify config.

### [1.3](https://github.com/IQTLabs/pipette/compare/v1.2...v1.3) (2021-01-14)

Fix shell var quoting
Fix Docker build
Upgrade pytype, eventlet, codecov-action

### [1.2](https://github.com/IQTLabs/pipette/compare/v1.1...v1.2) (2020-08-06)

Repo move, dependency and documentation updates.

### [1.1](https://github.com/IQTLabs/pipette/compare/v1.0...v1.1) (2020-03-06)

Fix for CLI space handling.

### [1.0](https://github.com/IQTLabs/pipette/compare/a550f8d...v1.0) (2020-02-27)

Initial version with IPv4/6 TCP/UDP support.


test_pipette.py
#!/usr/bin/env python

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import ipaddress
import socket
import unittest
from pipette import FAKEPORT
from pipette import Pipette

from os_ken.lib.packet import ethernet, icmpv6, ipv6, packet, vlan
from os_ken.ofproto import ether


class FakePort:

    port_no = 1
    name = 'fakey'
    state = 0


class FakeDP:

    def __init__(self):
        self.ports = {1: FakePort()}
        self.msgs = []

    def send_msg(self, msg):
        self.msgs.append(msg)


class FakeEv:

    def __init__(self):
        self.dp = FakeDP()



class PipetteSmokeTest(unittest.TestCase):  # pytype: disable=module-attr
    """Test bare instantiation of controller classes."""



    def test_smoke_connect(self):
        for nfvip in ('192.168.1.1/16', 'fc00::1/64'):
            pipette = Pipette(dpset={})
            pipette.NFVIPS = [ipaddress.ip_interface(nfvip)]
            assert pipette.dp_connect(FakeEv()) is None


    def test_smoke_packet_in(self):
        nd_solicit = packet.Packet()
        eth_src = '01:02:03:04:05:06'
        eth_dst = 'ff:ff:ff:ff:ff:ff'
        src_ip = 'fc00::1'
        dst_ip = 'fc00::2'
        vid = 2
        for protocol in (
                ethernet.ethernet(eth_dst, eth_src, ether.ETH_TYPE_8021Q),
                vlan.vlan(vid=vid, ethertype=ether.ETH_TYPE_IPV6),
                ipv6.ipv6(src=src_ip, dst=dst_ip, nxt=socket.IPPROTO_ICMPV6, hop_limit=255),
                icmpv6.icmpv6(
                    type_=icmpv6.ND_NEIGHBOR_SOLICIT,
                    data=icmpv6.nd_neighbor(dst=src_ip, option=icmpv6.nd_option_tla(hw_src=eth_src), res=7))):
            nd_solicit.add_protocol(protocol)
        nd_solicit.serialize()


        fake_dp = FakeDP()
        fake_pipette = Pipette(dpset={})

        class FakeMsg:

            def __init__(self):
                self.datapath = fake_dp
                self.match = {'in_port': FAKEPORT}
                self.data = nd_solicit.data


        class FakePiEv:

            def __init__(self):
                self.msg = FakeMsg()


        fake_pipette = Pipette(dpset={})
        fake_pipette.packet_in_handler(FakePiEv())
        assert fake_dp.msgs



if __name__ == "__main__":
    unittest.main()  # pytype: disable=module-attr


test_ovs_pipette.sh
#!/bin/bash

# requires OVS and Ryu
# Test that NAT works.

set -e

OF=6699
TMPDIR=$(mktemp -d)
SANDBOX=$TMPDIR/ovs-sandbox.sh
TESTSCRIPT=$TMPDIR/test-script.sh
TESTOUTPUT=$TMPDIR/test-output.txt

wget -q -O$SANDBOX https://raw.githubusercontent.com/openvswitch/ovs/master/tutorial/ovs-sandbox
chmod +x $SANDBOX

cat > $TESTSCRIPT <<- EOTESTSCRIPT
ovs-vsctl add-br copro0 \
         -- set bridge copro0 other-config:datapath-id=0000000000000001 \
         -- add-port copro0 copro -- set interface copro ofport_request=1 \
         -- add-port copro0 fake -- set interface fake ofport_request=2 \
         -- set-controller copro0 tcp:127.0.0.1:$OF \
         -- set controller copro0 connection-mode=out-of-band

while [ "\$(ovs-vsctl show|grep -i 'is_connected: true')" = "" ] ; do
  sleep 1
  echo .
done

ovs-appctl ofproto/trace copro0 in_port=copro,dl_vlan=2,dl_src=0e:00:00:00:00:01,dl_dst=0e:00:00:00:00:02,eth_type=0x800,nw_src=192.168.2.5,nw_dst=192.168.2.1,nw_proto=6,tcp_src=9999,tcp_dst=80 -generate

ovs-appctl ofproto/trace copro0 in_port=fake,dl_vlan=2,dl_src=0e:00:00:00:00:66,dl_dst=0e:00:00:00:00:67,eth_type=0x800,nw_src=10.10.0.1,nw_dst=10.10.5.1,nw_proto=6,tcp_src=80,tcp_dst=9999 -generate
EOTESTSCRIPT

osken-manager pipette.py --verbose --ofp-tcp-listen-port $OF &
RPID=$!

chmod +x $TESTSCRIPT
SHELL=$TESTSCRIPT $SANDBOX > $TESTOUTPUT
kill $RPID

grep "Final flow" $TESTOUTPUT

# inbound NAT entry
grep -q "Final flow: tcp,reg1=0x2,reg2=0x1,reg7=0xa0a0501,in_port=1,dl_vlan=2,dl_vlan_pcp=0,vlan_tci1=0x0000,dl_src=0e:00:00:01:00:02,dl_dst=0e:00:00:01:00:01,nw_src=10.10.5.1,nw_dst=10.10.0.1,nw_tos=0,nw_ecn=0,nw_ttl=0,tp_src=9999,tp_dst=80,tcp_flags=0" $TESTOUTPUT

# outbound NAT entry
grep -q "Final flow: tcp,in_port=2,dl_vlan=2,dl_vlan_pcp=0,vlan_tci1=0x0000,dl_src=0e:00:00:00:00:02,dl_dst=0e:00:00:00:00:01,nw_src=192.168.2.1,nw_dst=192.168.2.5,nw_tos=0,nw_ecn=0,nw_ttl=0,tp_src=80,tp_dst=9999,tcp_flags=0" $TESTOUTPUT

rm -rf $TMPDIR

echo PASS


runpipette.sh
#!/bin/bash

source ./pipetteconf.sh

function show_help()
{
    echo "pipette coprocessor setup (uses sudo)

    Usage: runpipette [option]
    Options:
      -c,  coproint      interface to send coprocessed traffic to
      -f,  fakeint       interface created for fake services to run on
      -m,  fakemac       fake mac for fake interface
      -i,  fakeips       fake ip for fake services, space delimitted (will be proxied from real IPS)
      -h,  help          print this help
      -b,  bridge        name of ovs bridge to create
      -p,  port          pipette port
      -v,  vlans         coprocessor vlans, space delimitted
      -r,  record        record traffic captured by pipette should be followed by location then size of file i.e.: -r /pcaps.file.pcap 50"
}

function check_args()
{
    while [ $# -gt 1 ]; do
        case $1 in
            -c|coproint)
                COPROINT="$2"
                shift
                ;;
            -f|fakeint)
                FAKEINT="$2"
                shift
                ;;
            -m|fakemac)
                FAKESERVERMAC="$2"
                shift
                ;;
            -i|fakeips)
                NFVIPS=""
                while [[ "$2" != -* && -n "$2" ]]
                do
                  NFVIPS+="$2"
                  if [[ "$3" != -* && -n "$3" ]]; then
                    NFVIPS+=" "
                  fi
                  shift
                done 
                ;;
            -v|vlans)
                VLANS=""
                while [[ "$2" != -* && -n "$2" ]]
                do
                  VLANS+="$2"
                  if [[ "$3" != -* && -n "$3" ]]; then
                    NFVIPS+=" "
                  fi
                  shift
                done 
                ;;
            -b|bridge)
                BR="$2"
                shift
                ;;
            -p|port)
                OF="$2"
                shift
                ;;
            -r|record)
                RECORD=1
                PCAP_LOCATION="$2"
                shift
                FILE_SIZE="$2"
                shift
                ;;
            -h|\?|help)
                show_help
                exit
                ;;
        esac
        shift
    done
}

if [ $# -gt 0 ]; then
    check_args "$@"
fi

if [ ! -d "$PIPETTE_TEMP_DIR" ]; then
  mkdir "$PIPETTE_TEMP_DIR"
fi

export BR VLANS COPROINT FAKEINT COPROPORT FAKEPORT FAKESERVERMAC NFVIPS OF
sudo --preserve-env="BR,VLANS,COPROINT,FAKEINT,COPROPORT,FAKEPORT,FAKESERVERMAC,NFVIPS,OF" ./configureovs.sh || exit 1

if [ $RECORD -ne 0 ]; then
    echo "Starting tcpdump on interface $COPROINT"
    sudo tcpdump -i "$COPROINT" -w "$PCAP_LOCATION" -C "$FILE_SIZE" -Z root &
    echo $! >> "$PIPETTE_TEMP_DIR/tcpdump"
fi

ryu-manager --verbose --ofp-tcp-listen-port "$OF" pipette.py &
echo $! >> "$PIPETTE_TEMP_DIR/ryu"


configureovs.sh
#!/bin/bash

while ovs-vsctl show; [ $? -ne 0 ]; do
  sleep 1
  echo waiting for OVS
done

set -e

OLDBR=copro${ID}
BR=pipette${ID}
FAKEINT=fake${ID}
FAKESERVERMAC=${FAKEMACPREFIX}:${ID}:${COPROPORT}
FAKECLIENTMAC=${FAKEMACPREFIX}:${ID}:${FAKEPORT}

# Configure pipette's OVS switch.
echo "Configuring OVS bridge $BR for pipette ID ${ID}"

remove_int_ip() {
  local int="$1"
  ip addr flush dev "$int"
  ip link set "$int" up
}

ovs-vsctl --if-exists del-br "$OLDBR"
ovs-vsctl --if-exists del-br "$BR"
echo "Configuring bridge"
ovs-vsctl add-br "$BR"
echo "Adding ports"
ovs-vsctl add-port "$BR" "$COPROINT" -- set Interface "$COPROINT" ofport_request="$COPROPORT"
ovs-vsctl add-port "$BR" "$FAKEINT" -- set Interface "$FAKEINT" ofport_request="$FAKEPORT" type=internal
ovs-ofctl del-flows "$BR"
echo "Setting controller"
ovs-vsctl set-controller "$BR" tcp:127.0.0.1:"$OF"

for i in $COPROINT $FAKEINT $BR ; do
  remove_int_ip "$i"
done

for ((i=0; i< "${#VLANS[@]}"; i++)) ; do
  vlan="${VLANS[$i]}"
  nfvip="${NFVIPS[$i]}"
  fakeintvlan=${FAKEINT}.${vlan}
  ip link add link "$FAKEINT" name "$fakeintvlan" type vlan id "$vlan"
  ip link set dev "$fakeintvlan" address "$FAKESERVERMAC"
  remove_int_ip "$fakeintvlan"
  ip addr add "$nfvip" dev "$fakeintvlan"
done

flows=""
while [[ "$flows" == "" ]] ; do
  flows=$(ovs-ofctl dump-flows $BR table=2|grep cookie|cat)
  sleep 1
  echo waiting for flows in $BR
done

IFS=, read -ra tcmap <<< "$TC"
for tc in "${tcmap[@]}" ; do
  IFS=: read -ra tcentry <<< "$tc"
  invid=${tcentry[0]}
  outvid=${tcentry[1]}
  tcpol=${tcentry[2]}
  suffix=${ID}v${invid}
  tcint=tc${suffix}
  tcbr=tcbr${suffix}
  echo configure tc map, input vid $invid, output vid $outvid, via $tcint/$tcbr
  ovs-vsctl add-port "$BR" "$tcint" -- set Interface "$tcint" ofport_request="$invid" type=internal
  ovs-ofctl add-flow "$BR" priority=64738,in_port=$invid,actions=output:$COPROPORT
  ovs-ofctl add-flow "$BR" priority=64738,in_port=$COPROPORT,dl_vlan=$invid,actions=mod_vlan_vid=$outvid,output:$invid
  ip link del $tcbr 2> /dev/null || true
  ip link add name $tcbr type bridge
  ip link set dev $tcbr up
  ip link set dev $tcint up
  ip link set dev $tcint master $tcbr
  bridge link set dev $tcint hairpin on
  tc qdisc add dev $tcint $tcpol
done

echo configuration successful


shutdownpipette.sh
#!/bin/bash

source ./pipetteconf.sh

function show_help()
{
    echo "pipette coprocessor shutdown (uses sudo)

    Usage: shutdownpipette [option]
    Options:
      -c,  coproint      interface to send coprocessed traffic to
      -f,  fakeint       interface created for fake services to run on
      -h,  help          print this help
      -b,  bridge        name of ovs bridge to create"
}

function check_args()
{
    while [ $# -gt 1 ]; do
        case $1 in
            -c|coproint)
                COPROINT="$2"
                shift
                ;;
            -f|fakeint)
                FAKEINT="$2"
                shift
                ;;
            -b|bridge)
                BR="$2"
                shift
                ;;
            -h|\?|help)
                show_help
                exit
                ;;
        esac
        shift
    done
}

if [ $# -gt 0 ]; then
    check_args "$@"
fi

if [ -f "$PIPETTE_TEMP_DIR/ryu" ]; then
    ryu_pid=$(cat "$PIPETTE_TEMP_DIR/ryu")
    echo "killing process with pid $ryu_pid"
    sudo kill -9 $ryu_pid
fi

if [ -f "$PIPETTE_TEMP_DIR/tcpdump" ]; then
    tcpdump_pid=$(cat "$PIPETTE_TEMP_DIR/tcpdump")
    echo "killing process with pid $tcpdump_pid"
    sudo kill -9 $tcpdump_pid
fi

#delete bridge
sudo ovs-vsctl del-br "$BR"

#reset coprocessor interface
sudo ip link set "$COPROINT" down
sudo ip link set "$COPROINT" up

if [ -d "$PIPETTE_TEMP_DIR" ]; then
  rm -rf "$PIPETTE_TEMP_DIR"
fi


pipette.py
#!/usr/bin/python3

# pipette implements a simple TCP-only L2/L3 proxy between a
# real broadcast domain/subnet and a fake one.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import ipaddress
import logging
import socket
import os
import sys
import netaddr
from os_ken.base import app_manager
from os_ken.controller import dpset, ofp_event
from os_ken.controller.handler import MAIN_DISPATCHER, set_ev_cls
from os_ken.lib.packet import arp, ethernet, icmpv6, ipv6, packet, vlan
from os_ken.ofproto import ether, nicira_ext
from os_ken.ofproto import ofproto_v1_3 as ofp
from os_ken.ofproto import ofproto_v1_3_parser as parser

# OVS port facing coprocessor (expects packets with a tag from the configured VLAN/s)
# Coprocessor only accepts packets with a VLAN tag.
COPROPORT = int(os.getenv('COPROPORT', '1'))
# OVS port facing fake services.
FAKEPORT = int(os.getenv('FAKEPORT', '2'))
# Fake interface must have this MAC.
ID = int(os.getenv('ID', 0))
FAKE_MAC_PREFIX = os.getenv('FAKEMACPREFIX', '0e:00:00:01')
FAKESERVERMAC = netaddr.EUI(FAKE_MAC_PREFIX + (':%u:%u' % (ID, COPROPORT)), dialect=netaddr.mac_unix)
# We will fake all coprocessed hosts as having this MAC.
FAKECLIENTMAC = netaddr.EUI(FAKE_MAC_PREFIX + (':%u:%u' % (ID, FAKEPORT)), dialect=netaddr.mac_unix)
# VLAN(s) to coprocess
VLANS = [int(vlan) for vlan in os.getenv('VLANS', '2').strip().split(' ')]
# IP addresses of fake services.
NFVIPS = [ipaddress.ip_interface(nfvip) for nfvip in os.getenv('NFVIPS', '10.10.0.1/16').strip().split(' ')]


class Pipette(app_manager.OSKenApp):

    # Idle timeout for translated flows (garbage collect)
    IDLE = 300

    INTF_TABLE = 0
    FROM_COPRO_TABLE = 1
    TO_COPRO_TABLE = 2
    AREG = 'xxreg1'
    FAKEPORTREG = 'reg1'
    COPROPORTREG = 'reg2'

    OFP_VERSIONS = [ofp.OFP_VERSION]
    _CONTEXTS = {
        'dpset': dpset.DPSet,
    }


    @staticmethod
    def send_mods(datapath, mods):
        for mod in mods:
            datapath.send_msg(mod)


    @staticmethod
    def apply_actions(actions):
        return [parser.OFPInstructionActions(ofp.OFPIT_APPLY_ACTIONS, actions)]


    @staticmethod
    def reg_copy(areg, dst, n_bits):
        # TODO: NXFlowSpecLoad/NXRegLoad doesn't work for > 64 bits.
        reg_bits = 64
        if n_bits < reg_bits:
            reg_bits = n_bits
        # pylint: disable=no-member
        return [
            parser.NXFlowSpecLoad(src=(areg, i*reg_bits), dst=(dst, i*reg_bits), n_bits=reg_bits)
            for i in range(int(n_bits / reg_bits))]


    def nat_actions(self, eth_type, nfvip, nat_offset):
        ip_ver = nfvip.ip.version
        ip_src_nxm = 'ipv%u_src_nxm' % ip_ver
        ip_dst_nxm = 'ipv%u_dst_nxm' % ip_ver
        ipbits = nfvip.ip.max_prefixlen
        # pylint: disable=no-member
        return [
            parser.NXActionRegMove(src_field='ipv%u_src' % ip_ver, dst_field=self.AREG, n_bits=nat_offset, src_ofs=0, dst_ofs=nat_offset),
            parser.NXActionRegMove(src_field='ipv%u_dst' % ip_ver, dst_field=self.AREG, n_bits=nat_offset, src_ofs=0, dst_ofs=0),
            # we have to load output port numbers into reg1 and reg2 because NXFlowSpecOutput() won't take a literal.
            parser.NXActionRegLoad(value=FAKEPORT, dst=self.FAKEPORTREG, ofs_nbits=nicira_ext.ofs_nbits(0, 15)),
            parser.NXActionRegLoad(value=COPROPORT, dst=self.COPROPORTREG, ofs_nbits=nicira_ext.ofs_nbits(0, 15)),
            # now program an inbound flow to perform NAT.
            parser.NXActionLearn(
                table_id=self.FROM_COPRO_TABLE,
                priority=2,
                hard_timeout=self.IDLE,
                specs=[
                    parser.NXFlowSpecMatch(src=eth_type, dst=('eth_type_nxm', 0), n_bits=16),
                    parser.NXFlowSpecMatch(src=(ip_src_nxm, 0), dst=(ip_src_nxm, 0), n_bits=ipbits),
                    parser.NXFlowSpecMatch(src=(ip_dst_nxm, 0), dst=(ip_dst_nxm, 0), n_bits=ipbits),
                    parser.NXFlowSpecLoad(src=int(FAKECLIENTMAC), dst=('eth_src_nxm', 0), n_bits=48),
                    parser.NXFlowSpecLoad(src=int(FAKESERVERMAC), dst=('eth_dst_nxm', 0), n_bits=48),
                ] + self.reg_copy(self.AREG, ip_src_nxm, ipbits) + [
                    parser.NXFlowSpecLoad(src=int(nfvip.ip), dst=(ip_dst_nxm, 0), n_bits=ipbits),
                    parser.NXFlowSpecOutput(src=(self.FAKEPORTREG, 0), dst='', n_bits=16),
                ]),
            # now program outbound an outbound flow.
            parser.NXActionLearn(
                table_id=self.TO_COPRO_TABLE,
                priority=2,
                idle_timeout=self.IDLE,
                specs=[
                    parser.NXFlowSpecMatch(src=eth_type, dst=('eth_type_nxm', 0), n_bits=16),
                    parser.NXFlowSpecMatch(src=int(nfvip.ip), dst=(ip_src_nxm, 0), n_bits=ipbits),
                ] + self.reg_copy(self.AREG, ip_dst_nxm, ipbits) + [
                    parser.NXFlowSpecLoad(src=('eth_dst_nxm', 0), dst=('eth_src_nxm', 0), n_bits=48),
                    parser.NXFlowSpecLoad(src=('eth_src_nxm', 0), dst=('eth_dst_nxm', 0), n_bits=48),
                    parser.NXFlowSpecLoad(src=(ip_dst_nxm, 0), dst=(ip_src_nxm, 0), n_bits=ipbits),
                    parser.NXFlowSpecLoad(src=(ip_src_nxm, 0), dst=(ip_dst_nxm, 0), n_bits=ipbits),
                    parser.NXFlowSpecOutput(src=(self.COPROPORTREG, 0), dst='', n_bits=16),
                ]),
            # now that future flows are programmed, handle the packet we have.
            parser.OFPActionSetField(eth_src=FAKECLIENTMAC),
            parser.OFPActionSetField(eth_dst=FAKESERVERMAC),
            parser.NXActionRegMove(src_field=self.AREG, dst_field=('ipv%u_src' % ip_ver), n_bits=ipbits, src_ofs=0, dst_ofs=0),
            parser.OFPActionSetField(**{'ipv%u_dst' % ip_ver: str(nfvip.ip)}),
            parser.OFPActionOutput(FAKEPORT)
        ]


    def natv6_flows(self, nfvip):
        nat_base = nfvip.network.network_address
        assert nfvip.network.prefixlen == 64, 'NFVIPS IPv4 all must be /64'
        # pylint: disable=no-member
        return self.apply_actions([
            # ipv6_src fc01::5 -> fc04::5:0:1, ipv6_dst fc01::1 -> fc04::1
            parser.NXActionRegLoad(value=(int(nat_base) & ((2**64)-1)), dst=self.AREG, ofs_nbits=nicira_ext.ofs_nbits(0, 63)),
            parser.NXActionRegLoad(value=(int(nat_base) >> 64), dst=self.AREG, ofs_nbits=nicira_ext.ofs_nbits(64, 127)),
        ] + self.nat_actions(ether.ETH_TYPE_IPV6, nfvip, 32))


    def natv4_flows(self, nfvip):
        nat_base = nfvip.network.network_address
        assert nfvip.network.prefixlen == 16, 'NFVIPS IPv4 all must be /16'
        # pylint: disable=no-member
        return self.apply_actions([
            # ipv4_src 192.168.2.5->10.10.5.1, ipv4_dst 192.168.2.1->10.10.0.1
            parser.NXActionRegLoad(value=int(nat_base), dst=self.AREG, ofs_nbits=nicira_ext.ofs_nbits(0, 31)),
        ] + self.nat_actions(ether.ETH_TYPE_IP, nfvip, 8))


    def common_reply_actions(self):
        # pylint: disable=no-member
        return [
            parser.NXActionRegMove(src_field='eth_src', dst_field='eth_dst', n_bits=48, src_ofs=0, dst_ofs=0),
            parser.OFPActionSetField(eth_src=FAKECLIENTMAC),
            parser.OFPActionOutput(ofp.OFPP_IN_PORT)
        ]


    def arp_reply_actions(self):
        # pylint: disable=no-member
        common_reply = self.common_reply_actions()
        return self.apply_actions([
            parser.NXActionRegLoad(value=arp.ARP_REPLY, dst='arp_op', ofs_nbits=nicira_ext.ofs_nbits(0, 2)),
            parser.NXActionRegMove(src_field='arp_sha', dst_field='arp_tha', n_bits=48, src_ofs=0, dst_ofs=0),
            parser.NXActionRegMove(src_field='arp_tpa', dst_field=self.AREG, n_bits=32, src_ofs=0, dst_ofs=0),
            parser.NXActionRegMove(src_field='arp_spa', dst_field='arp_tpa', n_bits=32, src_ofs=0, dst_ofs=0),
            parser.NXActionRegMove(src_field=self.AREG, dst_field='arp_spa', n_bits=32, src_ofs=0, dst_ofs=0),
            parser.OFPActionSetField(arp_sha=FAKECLIENTMAC),
            ] + common_reply)


    def tcp_udp_flows(self, vlan_id, nfvip, eth_type, nat_flows):
        return (
            # Learn from coprocessor port/do inbound translation.
            (self.FROM_COPRO_TABLE, parser.OFPMatch(eth_type=eth_type, vlan_vid=vlan_id),
             nat_flows(nfvip)),
            # Packets from coprocessor go to tuple inbound table.
            (self.INTF_TABLE, parser.OFPMatch(
                eth_type=eth_type, ip_proto=socket.IPPROTO_TCP, in_port=COPROPORT, vlan_vid=vlan_id),
             [parser.OFPInstructionGotoTable(self.FROM_COPRO_TABLE)]),
            (self.INTF_TABLE, parser.OFPMatch(
                eth_type=eth_type, ip_proto=socket.IPPROTO_UDP, in_port=COPROPORT, vlan_vid=vlan_id),
             [parser.OFPInstructionGotoTable(self.FROM_COPRO_TABLE)]),
            # Packets from fake interface go outbound table.
            (self.INTF_TABLE, parser.OFPMatch(
                eth_type=eth_type, ip_proto=socket.IPPROTO_TCP, in_port=FAKEPORT, vlan_vid=vlan_id),
             [parser.OFPInstructionGotoTable(self.TO_COPRO_TABLE)]),
            (self.INTF_TABLE, parser.OFPMatch(
                eth_type=eth_type, ip_proto=socket.IPPROTO_UDP, in_port=FAKEPORT, vlan_vid=vlan_id),
             [parser.OFPInstructionGotoTable(self.TO_COPRO_TABLE)]))


    def ipv6_flows(self, vlan_id, nfvip):
        return (
            (self.INTF_TABLE, parser.OFPMatch(
                eth_type=ether.ETH_TYPE_IPV6, vlan_vid=vlan_id, ip_proto=socket.IPPROTO_ICMPV6, icmpv6_type=icmpv6.ND_NEIGHBOR_SOLICIT),
             self.apply_actions(
                 [parser.OFPActionOutput(ofp.OFPP_CONTROLLER)])),) + self.tcp_udp_flows(vlan_id, nfvip, ether.ETH_TYPE_IPV6, self.natv6_flows)


    def ipv4_flows(self, vlan_id, nfvip):
        return (
            # Program OVS to respond to ARP on fake port.
            (self.TO_COPRO_TABLE, parser.OFPMatch(eth_type=ether.ETH_TYPE_ARP, vlan_vid=vlan_id),
             self.arp_reply_actions()),
            (self.INTF_TABLE, parser.OFPMatch(
                eth_type=ether.ETH_TYPE_ARP, eth_src=FAKESERVERMAC, in_port=FAKEPORT, arp_op=arp.ARP_REQUEST, vlan_vid=vlan_id),
             [parser.OFPInstructionGotoTable(self.TO_COPRO_TABLE)])) + self.tcp_udp_flows(vlan_id, nfvip, ether.ETH_TYPE_IP, self.natv4_flows)


    @set_ev_cls(dpset.EventDP, dpset.DPSET_EV_DISPATCHER)
    @set_ev_cls(dpset.EventDPReconnected, dpset.DPSET_EV_DISPATCHER)
    def dp_connect(self, ryu_event):
        logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
        datapath = ryu_event.dp
        for of_port in datapath.ports.values():
            self.report_port(of_port)
        mods = []
        # Drop all flows.
        mods.append(parser.OFPFlowMod(
            datapath=datapath,
            command=ofp.OFPFC_DELETE,
            out_port=ofp.OFPP_ANY,
            out_group=ofp.OFPG_ANY,
        ))
        # Default deny all tables
        for table_id in (self.INTF_TABLE, self.FROM_COPRO_TABLE, self.TO_COPRO_TABLE):
            mods.append(parser.OFPFlowMod(
                datapath=datapath,
                command=ofp.OFPFC_ADD,
                table_id=table_id,
                priority=0,
                instructions=[]))

        for nfvlan, nfvip in zip(VLANS, NFVIPS):
            vlan_id = (nfvlan | ofp.OFPVID_PRESENT)
            if nfvip.version == 6:
                flows = self.ipv6_flows
            else:
                flows = self.ipv4_flows
            for table_id, match, instructions in flows(vlan_id, nfvip):
                mods.append(parser.OFPFlowMod(
                    datapath=datapath,
                    command=ofp.OFPFC_ADD,
                    table_id=table_id,
                    priority=1,
                    match=match,
                    instructions=instructions))
        self.send_mods(datapath, mods)


    def report_port(self, of_port):
        port_names = {
            COPROPORT: 'COPROPORT',
            FAKEPORT: 'FAKEPORT',
        }
        port_name = port_names.get(of_port.port_no, None)
        if not port_name:
            return
        blocked_down_state = (
            (of_port.state & ofp.OFPPS_BLOCKED) or (of_port.state & ofp.OFPPS_LINK_DOWN))
        if blocked_down_state:
            port_state = 'down'
        else:
            port_state = 'up'
        logging.warning(
            'PORT %s (%u) %s: %s', of_port.name, of_port.port_no, port_state, of_port)


    @set_ev_cls(ofp_event.EventOFPPortStatus, MAIN_DISPATCHER) # pylint: disable=no-member
    def port_status_handler(self, ryu_event):
        self.report_port(ryu_event.msg.desc)


    # TODO: need packet in handler for IPv6 only, as OF1.3 won't let us set OFPXMT_OFB_ICMPV6_ND_RESERVED
    # See https://docs.opendaylight.org/projects/netvirt/en/latest/specs/fluorine/ovs_based_na_responder_for_gw.html.
    @set_ev_cls(ofp_event.EventOFPPacketIn, MAIN_DISPATCHER) # pylint: disable=no-member
    def packet_in_handler(self, event):
        if event.msg.match['in_port'] != FAKEPORT:
            return
        pkt = packet.Packet(event.msg.data)
        eth_protocol = pkt.get_protocol(ethernet.ethernet)
        vlan_protocol = pkt.get_protocol(vlan.vlan)
        ipv6_protocol = pkt.get_protocol(ipv6.ipv6)
        icmpv6_protocol = pkt.get_protocol(icmpv6.icmpv6)
        if not (eth_protocol and vlan_protocol and ipv6_protocol and icmpv6_protocol):
            return
        if icmpv6_protocol.type_ != icmpv6.ND_NEIGHBOR_SOLICIT:
            return
        if int(ipaddress.ip_address(ipv6_protocol.src)) == 0:
            return
        src_ip = ipaddress.ip_address(icmpv6_protocol.data.dst)
        if src_ip.is_reserved:
            return
        eth_dst = eth_protocol.src
        dst_ip = ipv6_protocol.src
        eth_src = FAKECLIENTMAC
        vid = vlan_protocol.vid
        reply = packet.Packet()
        for protocol in (
                ethernet.ethernet(eth_dst, eth_src, ether.ETH_TYPE_8021Q),
                vlan.vlan(vid=vid, ethertype=ether.ETH_TYPE_IPV6),
                ipv6.ipv6(src=src_ip, dst=dst_ip, nxt=socket.IPPROTO_ICMPV6, hop_limit=255),
                icmpv6.icmpv6(
                    type_=icmpv6.ND_NEIGHBOR_ADVERT,
                    data=icmpv6.nd_neighbor(dst=src_ip, option=icmpv6.nd_option_tla(hw_src=eth_src), res=7))):
            reply.add_protocol(protocol)
        reply.serialize()
        out = parser.OFPPacketOut(
            datapath=event.msg.datapath,
            buffer_id=ofp.OFP_NO_BUFFER,
            in_port=ofp.OFPP_CONTROLLER,
            actions=[parser.OFPActionOutput(FAKEPORT)], data=reply.data)
        self.send_mods(event.msg.datapath, [out])
